packages/ckeditor5-typing/src/input.ts

```diff
@@ -138,18 +195,276 @@ export default class Input extends Plugin {
+				if ( !view.document.isComposing ) {
+					return;
+				}
+
+				// Check if mutations are relevant for queued changes.
+				for ( const { node } of mutations ) {
+					const viewElement = findMappedViewAncestor( node, mapper );
+					const modelElement = mapper.toModelElement( viewElement )!;
+
+					if ( this._compositionQueue.isComposedElement( modelElement ) ) {
+						this._compositionQueue.flush( 'mutations' );
+
+						return;
+					}
+				}
+
+				// @if CK_DEBUG_TYPING // if ( ( window as any ).logCKETyping ) {
+				// @if CK_DEBUG_TYPING // 	console.log( '%c[Input]%c Mutations not related to the composition.',
+				// @if CK_DEBUG_TYPING // 		'font-weight: bold; color: green;', 'font-style: italic'
+				// @if CK_DEBUG_TYPING // 	);
+				// @if CK_DEBUG_TYPING // }
+			} );
+
+			// Make sure that all changes are applied to the model before the end of composition.
+			this.listenTo<ViewDocumentCompositionEndEvent>( view.document, 'compositionend', () => {
+				this._compositionQueue.flush( 'composition end' );
+			} );
+
+			// Trigger mutations check after the composition completes to fix all DOM changes that got ignored during composition.
+			this.listenTo<ViewDocumentCompositionEndEvent>( view.document, 'compositionend', () => {
+				const mutations: Array<MutationData> = [];
+
+				for ( const element of this._compositionQueue.flushComposedElements() ) {
+					const viewElement = mapper.toViewElement( element );
+
+					if ( !viewElement ) {
+						continue;
+					}
+
+					mutations.push( { type: 'children', node: viewElement } );
+				}
+
+				if ( mutations.length ) {
+					// @if CK_DEBUG_TYPING // if ( ( window as any ).logCKETyping ) {
+					// @if CK_DEBUG_TYPING // 	console.group( '%c[Input]%c Fire post-composition mutation fixes.',
+					// @if CK_DEBUG_TYPING // 		'font-weight: bold; color: green', 'font-weight: bold', ''
+					// @if CK_DEBUG_TYPING // 	);
+					// @if CK_DEBUG_TYPING // }
+
+					view.document.fire<ViewDocumentMutationsEvent>( 'mutations', { mutations } );
+
+					// @if CK_DEBUG_TYPING // if ( ( window as any ).logCKETyping ) {
+					// @if CK_DEBUG_TYPING // 	console.groupEnd();
+					// @if CK_DEBUG_TYPING // }
+				}
+			}, { priority: 'lowest' } );
+		} else {
+			// After composition end we need to verify if there are no left-overs.
+			// Listening at the lowest priority so after the InsertTextObserver added above.
+			this.listenTo<ViewDocumentCompositionEndEvent>( view.document, 'compositionend', () => {
+				// @if CK_DEBUG_TYPING // if ( ( window as any ).logCKETyping ) {
+				// @if CK_DEBUG_TYPING // 	console.group( '%c[Input]%c Force render after composition end.',
+				// @if CK_DEBUG_TYPING // 		'font-weight: bold; color: green', 'font-weight: bold', ''
+				// @if CK_DEBUG_TYPING // 	);
+				// @if CK_DEBUG_TYPING // }
+
+				view.document.fire<ViewDocumentMutationsEvent>( 'mutations', { mutations: [] } );
```

Comment:
Explain that this will trigger the renderer to check all the nodes previously marked but ignored because the document was in the composition state (which blocks the renderer).  
  
Also explain an example case when this is needed.

Reply:
Added in ce023f1.